app/backend/app/routes/query.py

```diff
@@ -0,0 +1,328 @@
+from uuid import uuid4
+
+import carnot
+import pandas as pd
+from fastapi import APIRouter, HTTPException
+from fastapi.responses import FileResponse, StreamingResponse
+from pydantic import BaseModel
+from sqlalchemy import select
+
+from app.database import (
+    AsyncSessionLocal,
+    Conversation,
+    Dataset,
+    DatasetFile,
+    Message,
+)
+
+router = APIRouter()
+logger = logging.getLogger(__name__)
+
+SESSION_TIMEOUT = timedelta(minutes=30)
+PROJECT_ROOT = Path(__file__).resolve().parents[4]
+BACKEND_ROOT = Path(__file__).resolve().parents[2]
+WEB_ROOT = Path(__file__).resolve().parents[3]
+active_sessions: Dict[str, Dict] = {}
+
+
+class QueryRequest(BaseModel):
+    query: str
+    dataset_ids: List[int]
+    session_id: Optional[str] = None
+
+
+def cleanup_old_sessions() -> None:
+    now = datetime.now()
+    expired_sessions = [
+        session_id
+        for session_id, session_data in active_sessions.items()
+        if now - session_data["last_access"] > SESSION_TIMEOUT
+    ]
+    for session_id in expired_sessions:
+        active_sessions.pop(session_id, None)
+
+
+async def get_or_create_conversation(
+    session_id: str, query: str, dataset_ids: List[int]
+) -> int:
+    async with AsyncSessionLocal() as db:
+        result = await db.execute(
+            select(Conversation).where(Conversation.session_id == session_id)
+        )
+        conversation = result.scalar_one_or_none()
+        if conversation:
+            return conversation.id
+
+        title = f"{query[:50]}..." if len(query) > 50 else query
+        dataset_ids_str = ",".join(map(str, dataset_ids))
+
+        conversation = Conversation(
+            session_id=session_id,
+            title=title,
+            dataset_ids=dataset_ids_str,
+        )
+        db.add(conversation)
+        await db.commit()
+        await db.refresh(conversation)
+        return conversation.id
+
+
+async def save_message(
+    conversation_id: int,
+    role: str,
+    content: str,
+    csv_file: Optional[str] = None,
+    row_count: Optional[int] = None,
+) -> None:
+    async with AsyncSessionLocal() as db:
+        message = Message(
+            conversation_id=conversation_id,
+            role=role,
+            content=content,
+            csv_file=csv_file,
+            row_count=row_count,
+        )
+        db.add(message)
+
+        result = await db.execute(
+            select(Conversation).where(Conversation.id == conversation_id)
+        )
+        conversation = result.scalar_one_or_none()
+        if conversation:
+            conversation.updated_at = datetime.utcnow()
+
+        await db.commit()
+
+
+def resolve_file_path(path: str) -> Optional[Path]:
+    candidates = [
+        Path(path),
+        PROJECT_ROOT / path,
+        WEB_ROOT / path,
+    ]
+    for candidate in candidates:
+        if candidate.exists():
+            return candidate
+    return None
+
+
+async def stream_query_execution(
```

Comment:
There's a lot happening in this function; it's fine as-is for now, but in the future we should break this down into smaller pieces